src/nix-env/nix-env.cc

```diff
@@ -911,36 +911,44 @@ static void queryJSON(Globals & globals, vector<DrvInfo> & elems, bool printOutP
 {
     JSONObject topObj(cout, true);
     for (auto & i : elems) {
-        JSONObject pkgObj = topObj.object(i.attrPath);
-
-        auto drvName = DrvName(i.queryName());
-        pkgObj.attr("name", drvName.fullName);
-        pkgObj.attr("pname", drvName.name);
-        pkgObj.attr("version", drvName.version);
-        pkgObj.attr("system", i.querySystem());
-
-        if (printOutPath) {
-            DrvInfo::Outputs outputs = i.queryOutputs();
-            JSONObject outputObj = pkgObj.object("outputs");
-            for (auto & j : outputs) {
-                outputObj.attr(j.first, j.second);
+        try {
+            if (i.hasFailed()) continue;
```

Comment:
I noticed this check for `hasFailed` also in the table output which I copied over.

It's a bit unfortunate all this functionality has to be copied for JSON/XML/Table...
If I was a bit more experienced in the codebase, I think creating a single intermediate representation and then having that serialize to XML/Table/JSON would be the best.

I'm sure that's pretty obvious assessment though :P